File: advall-python/d5.py
import hashlib

with open ("d5_input.txt", "r") as inputfile:
    input = inputfile.read()

pwd = ""
i = 0

while len(pwd) < 8:
    g = hashlib.md5()
    while g.hexdigest()[0:5] != "00000":
        g = hashlib.md5()
        g.update(input + str(i))
        i += 1
    pwd += str(g.hexdigest()[5])
print "Part one: " + str(pwd)

pwd = ['_', '_', '_', '_', '_', '_', '_', '_']
i = 0

while '_' in pwd:
    g = hashlib.md5()
    while g.hexdigest()[0:5] != "00000":
        g = hashlib.md5()
        g.update(input + str(i))
        i += 1
    if (int(g.hexdigest()[5], 16) < 8):
        if pwd[int(g.hexdigest()[5])] == '_':
            pwd[int(g.hexdigest()[5])] = g.hexdigest()[6]
print "Part two: " + "".join(pwd)

File: advall-haskell/d5.hs
--NOT DONE!!
import Data.Hash.MD5
import Data.List.Split
import Data.List

getKeyChar :: String -> String
getKeyChar s
    | take 5 s == "00000" = [head (drop 5 s)]
    | otherwise = ""

makeKey :: String -> Int -> String
makeKey doorID l = makeKey' getKeyChar doorID 0 "" l

makeKey' :: (String -> String) -> String -> Integer -> String -> Int -> String
makeKey' f doorID i key l
    | (length key) == l = key
    | otherwise = makeKey' f doorID (i+1) 
        (key ++ (getKeyChar (md5s (Str (doorID ++ (show i)))))) l

getKeyCharWithIndex :: String -> String
getKeyCharWithIndex s
    | (take 5 s == "00000") && ((read [head (drop 5 s)]) < 9) = take 2 (drop 5 s)
    | otherwise = ""

makeKey2 :: String -> Int -> [String]
--makeKey2 doorID l = map last (sort (chunksOf 2 (makeKey' getKeyCharWithIndex doorID 0 "" l)))
makeKey2 doorID l = (chunksOf 2 (makeKey' getKeyCharWithIndex doorID 0 "" l))

main = do
    input <- readFile "d5_input.txt"
--    let answer1 = makeKey input 8
 --   putStrLn $ "Part one: " ++ show answer1
    let answer2 = makeKey2 input 16
    putStrLn $ "Part two: " ++ show answer2